From c705a0b85d05e3b335dc572e2e0a000e1e7036e6 Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Mon, 18 Mar 2013 17:15:28 +0530
Subject: [PATCH] display failed submit conditions in Change Screen.

With this commit, Change Screen displays the list
of unsatisfied submit conditions which are defined
using 'condition' term in prolog.

This list is displayed above the ApprovalTable very
similar to missing labels which are being displayed
below ApprovalTable.

Change-Id: I7e2b1917421162cd80d753187243d9a0b7327850
---
 .../google/gerrit/client/changes/ChangeInfo.java   |   17 ++++
 .../google/gerrit/client/changes/ChangeScreen.java |    5 +
 .../client/changes/SubmitConditionBlock.java       |   78 ++++++++++++++++++++
 .../google/gerrit/server/change/ChangeJson.java    |   23 ++++++
 4 files changed, 123 insertions(+), 0 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/SubmitConditionBlock.java

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfo.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfo.java
index f4453da..afa371a 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfo.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeInfo.java
@@ -27,6 +27,7 @@ import com.google.gwt.core.client.JsArrayString;
 import com.google.gwtjsonrpc.client.impl.ser.JavaSqlTimestamp_JsonSerializer;
 
 import java.sql.Timestamp;
+import java.util.List;
 import java.util.Set;
 
 public class ChangeInfo extends JavaScriptObject {
@@ -34,6 +35,9 @@ public class ChangeInfo extends JavaScriptObject {
     if (labels0() != null) {
       labels0().copyKeysIntoChildren("_name");
     }
+    if (conditions0() != null) {
+      conditions0().copyKeysIntoChildren("_name");
+    }
   }
 
   public final Project.NameKey project_name_key() {
@@ -72,6 +76,10 @@ public class ChangeInfo extends JavaScriptObject {
     return labels0().keySet();
   }
 
+  public final Set<String> conditions() {
+    return conditions0().keySet();
+  }
+
   public final native String id() /*-{ return this.id; }-*/;
   public final native String project() /*-{ return this.project; }-*/;
   public final native String branch() /*-{ return this.branch; }-*/;
@@ -88,6 +96,8 @@ public class ChangeInfo extends JavaScriptObject {
   public final native String _sortkey() /*-{ return this._sortkey; }-*/;
   private final native NativeMap<LabelInfo> labels0() /*-{ return this.labels; }-*/;
   public final native LabelInfo label(String n) /*-{ return this.labels[n]; }-*/;
+  public final native NativeMap<ConditionInfo> conditions0() /*-{ return this.conditions; }-*/;
+  public final native ConditionInfo condition(String n) /*-{ return this.conditions[n]; }-*/;
 
   private final native NativeMap<JavaScriptObject> _permitted_labels()
   /*-{ return this.permitted_labels; }-*/;
@@ -107,6 +117,13 @@ public class ChangeInfo extends JavaScriptObject {
   protected ChangeInfo() {
   }
 
+  public static class ConditionInfo extends JavaScriptObject {
+    public final native SubmitRecord.Condition.Status status() /*-{ return this.status; }-*/;
+
+    protected ConditionInfo() {
+    }
+  }
+
   public static class LabelInfo extends JavaScriptObject {
     public final SubmitRecord.Label.Status status() {
       if (approved() != null) {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
index 15b033b..adf1aa4 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
@@ -66,6 +66,7 @@ public class ChangeScreen extends Screen
   private com.google.gerrit.client.changes.ChangeInfo changeInfo;
 
   private ChangeDescriptionBlock descriptionBlock;
+  private SubmitConditionBlock submitConditionBlock;
   private ApprovalTable approvals;
 
   private IncludedInTable includedInTable;
@@ -165,6 +166,9 @@ public class ChangeScreen extends Screen
     descriptionBlock = new ChangeDescriptionBlock(keysAction);
     add(descriptionBlock);
 
+    submitConditionBlock = new SubmitConditionBlock();
+    add(submitConditionBlock);
+
     approvals = new ApprovalTable();
     add(approvals);
 
@@ -294,6 +298,7 @@ public class ChangeScreen extends Screen
         detail.getAccounts(), detail.getSubmitTypeRecord());
     dependsOn.display(detail.getDependsOn());
     neededBy.display(detail.getNeededBy());
+    submitConditionBlock.display(changeInfo);
     approvals.display(changeInfo);
 
     patchesList.clear();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/SubmitConditionBlock.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/SubmitConditionBlock.java
new file mode 100644
index 0000000..5c83411
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/SubmitConditionBlock.java
@@ -0,0 +1,78 @@
+// Copyright (C) 2013 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.gerrit.client.changes;
+
+import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.client.ui.AccountLink;
+import com.google.gerrit.common.data.SubmitRecord;
+import com.google.gwt.user.client.DOM;
+import com.google.gwt.user.client.Element;
+import com.google.gwt.user.client.ui.Composite;
+import com.google.gwt.user.client.ui.Widget;
+
+import java.util.List;
+
+public class SubmitConditionBlock extends Composite {
+  private final Widget missing;
+  private ChangeInfo lastChange;
+
+  public SubmitConditionBlock() {
+    missing = new Widget() {
+      {
+        setElement(DOM.createElement("ul"));
+      }
+    };
+    missing.setStyleName(Gerrit.RESOURCES.css().missingApprovalList());
+
+    initWidget(missing);
+  }
+
+  void display(ChangeInfo change) {
+    lastChange = change;
+    removeAllChildren(missing.getElement());
+    for (String condition : change.conditions()) {
+      if (change.condition(condition).status() ==
+          SubmitRecord.Condition.Status.NEED) {
+        addMissingLabel(Util.M.needApproval(condition));
+      }
+    }
+
+    missing.setVisible(DOM.getChildCount(missing.getElement()) > 0);
+  }
+
+  private void removeAllChildren(Element el) {
+    for (int i = DOM.getChildCount(el) - 1; i >= 0; i--) {
+      DOM.removeChild(el, DOM.getChild(el, i));
+    }
+  }
+
+  private void addMissingLabel(String text) {
+    Element li = DOM.createElement("li");
+    li.setClassName(Gerrit.RESOURCES.css().missingApproval());
+    DOM.setInnerText(li, text);
+    DOM.appendChild(missing.getElement(), li);
+  }
+
+  private void reload() {
+    ChangeApi.detail(lastChange.legacy_id().get(),
+        new GerritCallback<ChangeInfo>() {
+          @Override
+          public void onSuccess(ChangeInfo result) {
+            display(result);
+          }
+        });
+  }
+}
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/change/ChangeJson.java b/gerrit-server/src/main/java/com/google/gerrit/server/change/ChangeJson.java
index 43ae87e..f2aa103 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/change/ChangeJson.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/change/ChangeJson.java
@@ -248,6 +248,7 @@ public class ChangeJson {
       out.permitted_labels = permittedLabels(cd);
       out.removable_reviewers = removableReviewers(cd, out.labels.values());
     }
+    out.conditions = conditionsFor(cd);
     out.finish();
 
     if (options.contains(ALL_REVISIONS) || options.contains(CURRENT_REVISION)
@@ -305,6 +306,22 @@ public class ChangeJson {
     return cd.getSubmitRecords();
   }
 
+  private Map<String, ConditionInfo> conditionsFor(ChangeData cd) throws OrmException {
+    Map<String, ConditionInfo> conditions =
+        new TreeMap<String, ConditionInfo>();
+    for (SubmitRecord rec : submitRecords(cd)) {
+      if (rec.conditions == null) {
+        continue;
+      }
+      for (SubmitRecord.Condition c : rec.conditions) {
+        ConditionInfo ci = new ConditionInfo();
+        ci.status = c.status;
+        conditions.put(c.name, ci);
+      }
+    }
+    return conditions;
+  }
+
   private Map<String, LabelInfo> labelsFor(ChangeData cd, boolean standard,
       boolean detailed) throws OrmException {
     if (!standard && !detailed) {
@@ -836,6 +853,8 @@ public class ChangeJson {
     Map<String, Collection<String>> permitted_labels;
     Collection<AccountInfo> removable_reviewers;
 
+    Map<String, ConditionInfo> conditions;
+
     String current_revision;
     Map<String, RevisionInfo> revisions;
     public Boolean _moreChanges;
@@ -891,6 +910,10 @@ public class ChangeJson {
     Integer linesDeleted;
   }
 
+  static class ConditionInfo {
+    SubmitRecord.Condition.Status status;
+  }
+
   static class LabelInfo {
     transient SubmitRecord.Label.Status _status;
 
-- 
1.7.8.3

